### ISSUE TITLE
Is there any way to cut flames make 60fps stickers not so slow after converted?

### ISSUE THREAD

**Issue by BrunoLiu** [ed-asriyan/tgs-to-gif#13]
As the title. Or is there any solutions to make the converted file play faster?

**Comment by bessx** [comment]
I am also having a problem with the output framerate.  After conversion, the gif seems to run in slow motion.

**Comment by ed-asriyan** [comment]
To check whether actual gif and tgs FPS values are different I've used the following python script from [agtoever's answer on Stackoverflow](https://stackoverflow.com/a/53365469):
```python
import sys
from PIL import Image

FILENAME = sys.argv[1]

def get_avg_fps(PIL_Image_object):
    """ Returns the average framerate of a PIL Image object """
    PIL_Image_object.seek(0)
    frames = duration = 0
    while True:
        try:
            frames += 1
            duration += PIL_Image_object.info['duration']
            PIL_Image_object.seek(PIL_Image_object.tell() + 1)
        except EOFError:
            return frames / duration * 1000
    return None

def main():
    img_obj = Image.open(FILENAME)
    print(f"Average fps: {get_avg_fps(img_obj)}")


if __name__ == '__main__':
    main()
```

and two generated gifs:
1. from https://t.me/addstickers/Mooncalfanimation (60 fps pack)
  ![a](https://user-images.githubusercontent.com/7848847/82101482-440c8500-9715-11ea-9d47-cd535f5cbfb4.gif)
2. from https://t.me/addstickers/SnappyCrab (30 fps pack)
  ![b](https://user-images.githubusercontent.com/7848847/82101492-4ec71a00-9715-11ea-8101-055614dd539e.gif)

Results:
1. On the first gif python script gives 59.78260869565218 FPS. Average duration of each frame is 30 -- 40 ms.
2. On the second gif python script gives 29.75206611570248 FPS. Average duration of each frame is 10 -- 20 ms.

So I conclude that tgs-to-gif **does not change the FPS** in result gif image.

Now I can only suggest changing the application for viewing GIFs. For example, the first gif on my phone opened by default gallery app displays normally, but in Google Photos app it displays "in slow motion".

**Comment by BrunoLiu** [comment]
It seems like the “slow motion” problem are probably caused by the viewer does not support 60fps gif rendering?
BTW I notices the title has been misspelled😂
This tool and telegram bot are very useful, thanks for working and replying.

**Comment by ed-asriyan** [comment]
> It seems like the “slow motion” problem are probably caused by the viewer does not support 60fps gif rendering?

Probably. My observations point to that. If you find out something else, please write here so that I and other people can solve the problem.


> This tool and telegram bot are very useful, thanks for working and replying.

Thanks for feedback :)

**Comment by bessx** [comment]
hmm, since I'm viewing them in google chrome, that seems odd.  Let me try again with those two sticker packs and see if I'm having the same issue.

**Comment by ed-asriyan** [comment]
I'm not a specialist in GIF format, but I found good explanation by Maadinsh why it is so: https://video.stackexchange.com/a/30160:

> There is no such thing as 60fps GIFs.
>
> GIF files store delay time between frames in whole hundredths of second, so it's impossible to make a GIF with exactly 60fps, because value 1 would give you 100fps and 2 would already have half of that - 50fps. Besides that, many programs (including Google Chrome) ignore value 1 and defaults to much slower animation speed when viewing such files, so 100fps GIFs, while technically possible according to file format, are not commonly used and probably a bad idea, so value 2 (50fps) is the most you can get.

**Comment by bessx** [comment]
![SnappyCrab-AgADoAADFkJrCg](https://user-images.githubusercontent.com/1159066/82106134-cfb9ed80-96e4-11ea-97f3-6fc56b22d64b.gif)

So Snappy Crab works beautifully and I can't test MoonCalfAnimation for some reason, but I did Lemurio Sticker Set:

![MrLemur-AgAD7QIAAvPjvgs](https://user-images.githubusercontent.com/1159066/82106202-2aebe000-96e5-11ea-8cc4-209181bc6a16.gif)

Plays slow in Windows Photo viewer, google chrome (like right on this page).

https://bestoftelegram.com/animated-stickers/MrLemur

**Comment by ed-asriyan** [comment]
@bessx Your second gif is slow in my browser too, but... I opened it in default gallery app on my phone (I simply downloaded image from your comment) and it is playing  faster than 60 fps (faster then in telegram).

**Comment by bessx** [comment]
Hmmmm, I see.  Then is there a way to tell the bot or set a flag to set it as you said, with 2 centiseconds per frame?  That way, even if it's slightly slower than 60fps, it will be ~50fps across on all devices?

**Comment by ed-asriyan** [comment]
It is not to simple because it uses third party library puppeteer-lottie that uses gifski tool... But is this the right decision anyway? It would be better and easier to create a tgs-to-mp4 bot.

**Comment by bessx** [comment]
I've been thinking about it. If the gif ends up at 10ms duration for each frame (maximum speed that causes compatibility issues), could the bot send back two versions of the gif? One with the 10ms duration and one with the next highest, 20 ms duration?

I think that would be a nice compromise without completely going for an mp4.  That way the user themselves can decide which one to use.

**Comment by bessx** [comment]
> It is not so simple because it uses third party library puppeteer-lottie that uses gifski tool... But is this the right decision anyway? It would be better and easier to create a tgs-to-mp4 bot.

I've been thinking about it. If the gif ends up at 10ms duration for each frame (maximum speed that causes compatibility issues), could the bot send back two versions of the gif? One with the 10ms duration and one with the next highest, 20 ms duration?

I think that would be a nice compromise without completely going for an mp4. That way the user themselves can decide which one to use.

**Comment by AndnixSH** [comment]
Also having the same issue with this sticker https://t.me/addstickers/Peepo_Pepe

**Comment by ed-asriyan** [comment]
I found pretty library [rlottie](https://github.com/Samsung/rlottie). In examples I found the program that converts lottie to gif. I tried it on sticker from https://github.com/ed-asriyan/tgs-to-gif/issues/13#issuecomment-629563728.
At first, I want to say it works much more faster, it is really cool. And resulting gifs have 50fps:
![out json](https://user-images.githubusercontent.com/7848847/82734259-87bd4b00-9d22-11ea-9d39-2a99c73af4aa.gif)

Script https://github.com/ed-asriyan/tgs-to-gif/issues/13#issuecomment-629542984 output:
```
python3 a.py out.json.gif 
Average fps: 50.0
```

But it does not generate transparent gifs, now I am trying to modify the lib to fix that. When I'll done I want to update the repo sources.

**Comment by AndnixSH** [comment]
Nice find!
I realized an online website called ezgif could adjust speed, by 150% it turned gif back to normal speed. I used the website to reduce size to be able to use them as discord emojies
https://ezgif.com/speed

![leaving](https://user-images.githubusercontent.com/40742924/82739256-7da94580-9d3e-11ea-8f3a-28526510070c.gif)

**Comment by ed-asriyan** [comment]
You can try new version that uses rlottie lib.
It stores in branch `master-cpp`. I also updated Dockerfile, so you can use it under docker container.

I made several tests. It have some artifacts as I see. For example, hat on the coala cap slightly changes its color during animation:
![tgs8 tgs](https://user-images.githubusercontent.com/7848847/82757370-ba317b80-9de8-11ea-8ca7-48367a55fc75.gif)

**Comment by bessx** [comment]
oh interesting, it's great that it work!

what's the duration per frame actually set to in these new gifs?

**Comment by ed-asriyan** [comment]
> what's the duration per frame actually set to in these new gifs?

2

> EDIT: It's beyond weird that the artifacting only happens when the background is transparent. I don't see a reason why that should happen in the transparent gif, but not in the one with a background.

[gif-h](https://github.com/charlietangora/gif-h) does not support transparency, it only has a PR https://github.com/charlietangora/gif-h/pull/7, which is not compiling (https://github.com/Evil-Spirit/gif-h/pull/1 should fix that) and has a color artefact, but it is better than nothing.

**Comment by ed-asriyan** [comment]
I patched puppeteer-lottie in https://github.com/ed-asriyan/puppeteer-lottie/commit/7bbe134d31b56bc33c71e5022a5d787589c6f25d and use it as a dependency in this project. So issue should be resolved.

**Comment by LouisWayne** [comment]
Tested the latest version with `npm`, but I still see slow gifs..

**Comment by ed-asriyan** [comment]
@LouisWayne,

Note that maximum fps for gifs is 50 (more details: https://github.com/ed-asriyan/tgs-to-gif/issues/13#issuecomment-629563654). If you sure, that fps is lower, please send me the link of the sticker to test.

**Comment by LouisWayne** [comment]
@ed-asriyan , thanks for reply!

Here is the download link: [https://gofile.io/d/aNsus5](https://gofile.io/d/aNsus5).  It would be great if you can find the solution for this sticker file :smile:

**Comment by ed-asriyan** [comment]
@LouisWayne, please, pull the latest master branch and reinstall nom dependencies (`rm -fr node_modules && npm I`). Problem should be resolved.

**Comment by LouisWayne** [comment]
@ed-asriyan , THANKS A LOT!

Now the generated gif file looks good! I hope the updates will apply to the Telegram bot soon :smile: You made an amazing library and I really appreciate it.  Thanks!

**Comment by ed-asriyan** [comment]
@LouisWayne, the update has already been applied to the Telegram bot. You're welcome!
